### PR TITLE
Avoid duration calculation panic when clock jumps back (#10544)

### PR DESCRIPTION
**Pull request by tabokie**
Cherry-pick #10544

### Release note 

```release-note
Fix duration calculation panics on certain platforms
```

### CODE REVIEW

**Review by yiwu-arbug — APPROVED** [review]

File: components/tikv_util/src/time.rs [hunk 1]
```diff
@@ -405,8 +454,10 @@ impl SubAssign<Duration> for Instant {
 impl Sub<Instant> for Instant {
     type Output = Duration;
 
+    // TODO: For safety in production code, `sub` actually does saturating_sub.
+    // We should remove this operator from public scope.
```
Comment (BusyJay) [inline components/tikv_util/src/time.rs:458]: What places is using `sub`? Is it safe to do `saturating_duration_since` for all of them?

File: components/tikv_util/src/time.rs [hunk 2]
```diff
@@ -317,12 +350,28 @@ impl Instant {
...
+        }
+    }
+
     // It is different from `elapsed_duration`, the resolution here is millisecond.
     // The processors in an SMP system do not start all at exactly the same time
     // and therefore the timer registers are typically running at an offset.
     // Use millisecond resolution for ignoring the error.
     // See more: https://linux.die.net/man/2/clock_gettime
-    fn elapsed_duration_coarse(later: Timespec, earlier: Timespec) -> Duration {
+    pub(crate) fn saturating_elapsed_duration_coarse(
```
Comment (BusyJay) [inline components/tikv_util/src/time.rs:371]: I don't think the rename makes sense as there is no accurate time when using coarse timer, so there is no such "non-saturating` version at all.

**Review by BusyJay — COMMENTED** [review]
I think we have a daemon thread to monitor jumping back of time. I can't say whether to panic is good or not as when monotonic timer is dumping back, there are many places are also affected including unify thread pool scheduling or hot fsm scheduling.

**Review by 5kbpers — APPROVED** [review]

File: components/tikv_util/src/time.rs (hunk 2 above)
Comment (tabokie) [reply components/tikv_util/src/time.rs:371]: make sense.

File: components/tikv_util/src/time.rs (hunk 1 above)
Comment (tabokie) [reply components/tikv_util/src/time.rs:455]: I replaced them in the last commit, PTAL.

**Comment by tabokie** [comment]
> I think we have a daemon thread to monitor jumping back of time. I can't say whether to panic is good or not as when monotonic timer is dumping back, there are many places are also affected including unify thread pool scheduling or hot fsm scheduling.

But I presume this won't affect service ability? Just quality. In this patch we will print error log to alert the admin to investigate if this can turn into a real problem.

File: components/tikv_util/src/time.rs (hunk 1 above)
Comment (BusyJay) [reply components/tikv_util/src/time.rs:455]: Seems the trait implementation can be deleted.

Comment (tabokie) [reply components/tikv_util/src/time.rs:455]: We are using it in "async_speed_limit::Limiter", I can make the op private though.

**Pull request merged**